tests/bluetooth/host/keys/bt_keys_get_addr/src/main.c

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/zephyr.h>
+#include "testing_common_defs.h"
+#include <addr.h>
+#include <keys.h>
+#include "mocks/keys_help_utils.h"
+
+/* This LUT contains different combinations of ID and Address pairs */
+const struct id_addr_pair testing_id_addr_pair_lut[CONFIG_BT_MAX_PAIRED] = {
+	{ BT_ADDR_ID_1, BT_ADDR_LE_1 },
+	{ BT_ADDR_ID_1, BT_ADDR_LE_2 },
+	{ BT_ADDR_ID_2, BT_ADDR_LE_1 },
+	{ BT_ADDR_ID_2, BT_ADDR_LE_2 }
+};
+
+/* Global iterator to iterate over the ID, Address pairs LUT */
+static uint8_t params_it;
+
+/* Pointer to the current set of testing parameters */
+struct id_addr_pair const *current_params_vector;
+
+/* This list will hold returned references while filling keys pool */
+struct bt_keys *returned_keys_refs[CONFIG_BT_MAX_PAIRED];
+
+BUILD_ASSERT(ARRAY_SIZE(testing_id_addr_pair_lut) == CONFIG_BT_MAX_PAIRED);
+BUILD_ASSERT(ARRAY_SIZE(testing_id_addr_pair_lut) == ARRAY_SIZE(returned_keys_refs));
+
+/*
+ *  Test list is empty
+ *
+ *  Constraints:
+ *   - Empty list at start up
+ *
+ *  Expected behaviour:
+ *   - List is empty at start up
+ */
+void test_list_is_empty_at_startup(void)
+{
+	zassert_true(check_key_pool_is_empty(),
+		     "List isn't empty, make sure to run this test just after a fresh start");
+}
+
+/*
+ *  Test filling the keys pool with (ID, Address) pairs
+ *
+ *  Constraints:
+ *   - Empty list at startup
+ *
+ *  Expected behaviour:
+ *   - A valid reference is returned by bt_keys_get_addr()
+ *   - ID value matches the one passed to bt_keys_get_addr()
+ *   - Address value matches the one passed to bt_keys_get_addr()
+ */
+void test_get_non_existing_keys(void)
+{
+	bt_addr_le_t addr;
+	struct bt_keys *returned_key;
+	uint8_t id = current_params_vector->id;
+
+	bt_addr_le_copy(&addr, current_params_vector->addr);
+
+	returned_key = bt_keys_get_addr(id, &addr);
+	/* 'params_it' is incremented before running the test case */
+	returned_keys_refs[params_it - 1] = returned_key;
+
+	zassert_true(returned_key != NULL, "bt_keys_get_addr() returned a non-valid reference");
+	zassert_true(returned_key->id == id,
+		     "bt_keys_get_addr() returned a reference with an incorrect ID");
+	zassert_true(!bt_addr_le_cmp(&returned_key->addr, &addr),
+		     "bt_keys_get_addr() set incorrect address %s value, expected %s",
+		     bt_addr_le_str(&returned_key->addr), bt_addr_le_str(&addr));
+	zassert_false(check_key_pool_is_empty(),
+		      "List is empty, (ID, Address) pair insertion failed");
+}
+
+/*
+ *  Test no equal references returned by bt_keys_get_addr()
+ *
+ *  Constraints:
+ *   - Keys pool has been filled
+ *
+ *  Expected behaviour:
+ *   - All returned references are different from each other
+ */
+void test_no_equal_references(void)
```

Comment:
I don't understand this test. You appear to be verifying that the existing key pool pointer does not have a duplicate in a list of pointers that are maintained locally and in the same order. That strikes me as a test of the behavior of the test suite rather than the key pool. The current implementation is an array, so it may be reasonable to verify that a returned pointer is NULL or within the range of the array (taking note of encapsulation concerns), but otherwise, testing the integrity of the array can be done with a controlled test to fill it up and check what it does and does not contain. The id has only 256 possible values (currently), so you can do a brute-force test against all ids.

I believe this test is also order-dependent, which is a concern.

Reply:
This test mainly verifies that an empty list is filled sequentially in a correct order, so that each reference pointer is moved correctly to the next location which also verifies that no reference pointer was returned twice.

The test iterates over a list with the same size as the keys pool, so boundaries are considered.

On the other hand, NULL checking is performed while adding new items through `bt_keys_get_addr()` (although the name is a bit confusing, this API tries to find a matched record with the same ID and address. If the item exists, it returns the item reference pointer. Otherwise, it tries to add the ID and address pair to the list and then returns the new record reference pointer)

Your concern that this test is order-dependent is totally valid, for example this test doesn't consider the situation when an item in the middle of the list is cleared then reused.